src/version_req.rs

```diff
@@ -232,7 +245,8 @@ impl VersionReq {
     /// assert!(exact.matches(&version));
     /// ```
     pub fn matches(&self, version: &Version) -> bool {
-        self.predicates.iter().all(|p| p.matches(version))
+        self.predicates.iter().all(|p| p.matches(version)) &&
+            self.predicates.iter().any(|p| p.pre_tag_is_compatible(version))
```

Comment:
ah ha! no, I think this is the bug. 

```
fn main() {
    let v: Vec<i32> = vec![];
    
    println!("any {}", v.iter().any(|_| true));
    println!("all {}", v.iter().all(|_| true));
}
```

prints `false` and then `true`, in other words, when `predicates` is empty, this is wrong.